Update plex-core to support synology packages

myellen opened the pull request:
Checks DISTRO variable in getPlexVersion and parseVersion to see if script is running on synology.

demonbane commented:
Is it even possible to install Plex from the command-line on a Synology NAS? I have one that I've got root SSH access on, but I don't know of any way to install the .spk packages from the command-line.

abl commented:
@demonbane `synopkg install <path to .spk file>` as root looks like it does the trick.

demonbane commented:
That's great! Ok, so this **can** be added because it is an officially supported platform from Plex, but there are a couple of issues that need to be addressed first. In addition to downloading the Synology package, this should also support:

* Installing the package
* Starting and stopping the package before/after installation if necessary (I'm not familiar with how synopkg works, so if it already does this automatically then it can be skipped)
* Verify that the script is running in a bash 4+ environment. DSM was shipping a 3.x bash version until recently, so some users might still have an old version of bash which won't work.
* Any additional logic specific to how Synology handles packages that I haven't listed here.

myellen commented:
Do you mean installing the package for the first time? or using your plexupdate installer script? It does handle installing the package using manual setup. I haven't modified installer.sh to detect the synology distro.

synopkg automatically starts and stops the package for installation.

I don't have an older version of dsm to test with. However, the changes I made don't use any bash functions that weren't already present in the script, so I'd assume it works fine in bash 3.x.

demonbane commented:
I did a bit of digging and it looks like most (all?) versions of DSM 6 already include bash 4 so I think that should be ok. The existing code is already pretty heavily dependent on bash 4 which is why I mentioned it.

For installation, it would be great if you update installer.sh to support initial setup as well, but the only requirement would be that plexupdate can auto-install the update after downloading it (presumably using the `synopkg` command provided above).

There should also be some logic added to try to determine if we're running on a synology automatically so that it doesn't have to be specified via an environment variable:

https://github.com/mrworf/plexupdate/blob/293e6d5ecbe6f41536a054fe91445c5cdd0d0c10/plexupdate.sh#L298-L315

I'm excited to see this added, so I'll try to be as responsive as possible to any questions you have so we can get this merged soon. Thanks for working on this!

myellen commented:
I patched plexupdate.sh to detect synology and to work with the new 5.json naming scheme.

I updated installer.sh too. Where does installer.sh set the BUILD variable? or is it not necessary? I define DISTRO and DISTRO_INSTALL inside of check_distro().

d1slact0r commented:
any update on this?

myellen commented:
@d1slact0r no. you can manually download [plexupdate-core](https://raw.githubusercontent.com/myellen/plexupdate/master/plexupdate-core) and [plexupdate.sh](https://raw.githubusercontent.com/myellen/plexupdate/master/plexupdate.sh) from my pull request and replace the versions you have on your synology now

deepfriedmind commented on plexupdate.sh:
```diff
@@ -307,6 +307,20 @@ if [ -z "${DISTRO_INSTALL}" ]; then
+		elif [ -f /etc/synoinfo.conf ]; then
+		        DISTRO="synology"
+		        if [ "${PUBLIC}" = "yes" ]; then
+			        if [ "${ARCH}" = "x86_64"]; then
```
These lines need spaces before the closing brackets, i.e. `[ "${ARCH}" = "x86_64" ]`.

myellen replied:
i added the spaces

BradleyMarie commented on plexupdate.sh:
```diff
@@ -307,6 +307,20 @@ if [ -z "${DISTRO_INSTALL}" ]; then
 				DISTRO_INSTALL="${REDHAT_INSTALL}"
 			fi
+		elif [ -f /etc/synoinfo.conf ]; then
+		        DISTRO="synology"
```
I noticed when I tried to copy this over the version from head. Some of the lines in this change were intended using both tabs and spaces.

myellen replied:
i changed the spaces to tabs

BradleyMarie commented on plexupdate.sh:
```diff
@@ -307,6 +307,20 @@ if [ -z "${DISTRO_INSTALL}" ]; then
+		elif [ -f /etc/synoinfo.conf ]; then
+		        DISTRO="synology"
+		        if [ "${PUBLIC}" = "yes" ]; then
+			        if [ "${ARCH}" = "x86_64" ]; then
```
I believe these build values are no longer accurate. This is what I see when I pull down the builds from Plex.

        "Synology": {
            "id": "synology",
            "name": "Synology",
            "release_date": 1557930938,
            "version": "1.15.6.1079-78232c603",
            "requirements": "Note: <b>Requires DSM 5.0 minimum to install.</b>",
            "extra_info": "Guide: <a href=\"https://support.plex.tv/hc/en-us/articles/205165858\" target=\"_blank\">Add Plex as a trusted publisher for package installations</a>.",
            "items_added": "(DVR) Add `Resolution`, `Replace lower resolution items`, and `Allow partial airings` at DVR level with option to override at recording level.",
            "items_fixed": "(Analysis) Corrected chapters not being imported again\r\n(DVR) Prefer HD airings when both HD and SD are available. (#9641)\r\nDocker host addresses were published unnecessarily (#9576)\r\nIP addresses from interfacing starting with a \"v\" weren'\''t published (#9576)\r\nSync or Optimize could fail to fallback to software transcoding if hardware transcoding failed (#9455)",
            "releases": [{
                "label": "Intel 32-bit (x10 Series, DS415play, and DS214play)",
                "build": "linux-x86",
                "distro": "synology",
                "url": "https://downloads.plex.tv/plex-media-server-new/1.15.6.1079-78232c603/synology/PlexMediaServer-1.15.6.1079-78232c603-x86.spk",
                "checksum": "1a6e5fceaf23031dab8518929c3dd8c2597fd93b"
            }, {
                "label": "Intel 64-bit (DSM 6.0 and newer)",
                "build": "linux-x86_64",
                "distro": "synology",
                "url": "https://downloads.plex.tv/plex-media-server-new/1.15.6.1079-78232c603/synology/PlexMediaServer-1.15.6.1079-78232c603-x86_64.spk",
                "checksum": "b1fb36ddce473b7518248e6ef67c1d01500fb72c"
            }, {
                "label": "ARMv8 (x18 Series)",
                "build": "linux-aarch64",
                "distro": "synology",
                "url": "https://downloads.plex.tv/plex-media-server-new/1.15.6.1079-78232c603/synology/PlexMediaServer-1.15.6.1079-78232c603-aarch64.spk",
                "checksum": "efc4caa3adc94347705b9e8483f0fd23c774d493"
            }, {
                "label": "ARMv7 (x13 Series, x14 Series (excluding DS4v14j), DS115j, RS815, and DS216se)",
                "build": "linux-armv7hf",
                "distro": "synology",
                "url": "https://downloads.plex.tv/plex-media-server-new/1.15.6.1079-78232c603/synology/PlexMediaServer-1.15.6.1079-78232c603-armv7hf.spk",
                "checksum": "7845bced023c11beb54e6a2b4cccf2bdbc20f55a"
            }, {
                "label": "ARMv7 (x15 Series (excluding DS115j and RS815), x16 Series (excluding DS216se), x17 Series, x18 Series, and DS414j)",
                "build": "linux-armv7hf_neon",
                "distro": "synology",
                "url": "https://downloads.plex.tv/plex-media-server-new/1.15.6.1079-78232c603/synology/PlexMediaServer-1.15.6.1079-78232c603-armv7hf_neon.spk",
                "checksum": "f3b85a650ce6c2bb7b5fe654e1758cf9c2877963"
            }]
        },

myellen replied:
Is that for public or private builds? When I added that stuff, only plexpass builds were using the new build values. Which is why if ${PUBLIC} is no, it uses
`else
				BUILD="linux-${ARCH}"`

If non-plexpass build values are now updated, I can remove those lines

BradleyMarie replied:
This was for public builds.

myellen replied:
just fixed it

BradleyMarie commented:
I believe you also need to update the AUTOSTART section to add handling for using synopkg to start the job after the update completes.

demonbane commented:
This fork is from an older release which still has the old API endpoints and build designations. It should be rebased to the current HEAD from upstream.

myellen commented:
> I believe you also need to update the AUTOSTART section to add handling for using synopkg to start the job after the update completes.

It's been working for me without that, but I'll double check

> This fork is from an older release which still has the old API endpoints and build designations. It should be rebased to the current HEAD from upstream.

I just rebased it.

demonbane commented on plexupdate-core:
```diff
@@ -228,13 +228,17 @@ isNewerVersion() {
 	fi
 }
 
 getPlexVersion() {
```
For readability, functions like these should generally start with the "default" option (Ubuntu/Debian) and then go on to the other cases. It's just a bit confusingly written right now because of how the `REDHAT` variable is used. I'd suggest swapping the two first cases and adding a comment (which I really should have done when I wrote the function 🤦‍♂ ), about what the `"${REDHAT}" != "yes"` does.)

demonbane replied:
The way GitHub is rendering this is a bit confusing. The comment is about the `getPlexVersion` function, not `isNewerVersion`.

myellen replied:
so if I just switch the cases like so
```
	if [ "${REDHAT}" != "yes" ]; then
		dpkg-query --showformat='${Version}' --show plexmediaserver 2>/dev/null
	elif [ "${DISTRO}" = "synology" ]; then
		synopkg version "Plex Media Server" 2>/dev/null
```
I don't think the switch ever reaches the synology case because ` "${REDHAT}" != "yes" ` is true for synology devices too.

I'm not sure how best to change it.

demonbane commented on plexupdate.sh:
```diff
@@ -307,6 +307,10 @@ if [ -z "${DISTRO_INSTALL}" ]; then
 			fi
+		elif [ -f /etc/synoinfo.conf ]; then
+			DISTRO="synology"
+			BUILD="linux-${ARCH}"
```
`linux-${ARCH}` is already set as the default in line 56, so this line can be removed.
```suggestion
```

demonbane commented:
Now that we're going to have 3 supported platforms, I think it's also time to retire the `REDHAT` variable altogether as well. If you'd like to take a stab at it I'd really appreciate it, but if not I can try to do it myself later.

myellen commented:
Before you install it on your synology, you'll have to follow the instructions in [How to add Plex’s package signing public key to Synology NAS Package Center](https://support.plex.tv/articles/205165858-how-to-add-plex-s-package-signing-public-key-to-synology-nas-package-center/). That link should be added to the readme or the wiki.


re: redhat, I don't have any linux boxes to test with, and I'm not very familiar with them anyways. So unfortunately, I'm not really comfortable making those changes.

myellen commented:
I added it to the README since you need to do that step whether you use automatic or manual setup.

demonbane commented:
Can the package signing key stuff be done via the command line? That would be a great thing to add to `installer.sh`.

myellen commented:
There's definitely no documented way to install the certificate. It's possible there's a hack-y way to do it. None of the programs I've checked in /usr/syno/bin/ so far can do it.

demonbane commented:
Ok, in that case we'll also need some code in both plexupdate and installer.sh that checks if synopkg threw an error related to the certificate and, if so, display a link to the Plex instructions for adding a key.

demonbane commented:
Just tested this on my synology and it partially worked. The package was installed, but it didn't automatically get started. I saw that `synopkg install` has a `--daemon-reload` option but I'm not sure if that'll do what we want or not.

Also, since it looks like DSM doesn't have a standard cron setup, installer.sh should probably just skip the cron setup parts altogether if DISTRO==synology.

BradleyMarie commented:
> Just tested this on my synology and it partially worked. The package was installed, but it didn't automatically get started. I saw that `synopkg install` has a `--daemon-reload` option but I'm not sure if that'll do what we want or not.

I believe this can be resolved by adding handing to the AUTOSTART section for Synology. I did this on my own box to resolve the same issue on my box.

demonbane commented:
Just tested this and it looks like `--daemon-reload` is only good for throwing an error and doesn't appear to do anything. I did try a new install as well as an upgrade, and in both cases the services is stopped and disabled but not automatically started afterwards.

Adding `synopkg start "Plex Media Server"` to the `AUTOSTART` logic should take care of both use cases.

myellen commented:
added an error message with instructions to go to the website and updated the autostart. I also hid the plexpass password when you type it in.

> Also, since it looks like DSM doesn't have a standard cron setup, installer.sh should probably just skip the cron setup parts altogether if DISTRO==synology.

When I run the installer, I just get the "It seems like you don't have a supported cron job setup, please see README.md for more details" message and the installation works fine. We could skip that check altogether, but my synology does have an /etc/crontab and /etc/cron.d that should work. However, I think the preferred method on synology should be to use the Task Scheduler in DSM. You'd only need a short page in the wiki to explain how to set it up.

demonbane commented on plexupdate-core:
```diff
@@ -51,7 +51,7 @@ getPlexToken() {
 		done
 		while true; do
-			read -e -p "PlexPass Password: " -i "$PASS" PASS
+			read -e -s -p "PlexPass Password: " -i "$PASS" PASS
```
I originally kept password echo on since plexupdate is meant to still be accessible to non-technical people, and they often get confused when no output appears while typing. (And bash still doesn't provide a way to replace characters with `*` or something similar.)

I'm ok with trying it and seeing if we get any issues filed about it, but I also view it as unnecessary in this case, so I'll leave the decision up to you.

demonbane commented on plexupdate.sh:
```diff
@@ -493,6 +493,9 @@ if [ "${AUTOINSTALL}" = "yes" ]; then
 		# Clarify why this failed, so user won't be left in the dark
 		error "Failed to install update. Command '${DISTRO_INSTALL} "${DOWNLOADDIR}/${FILENAME}"' returned error code ${RET}"
+		if [ "${DISTRO}" = "synology" -a ${RET} -eq 1 ]; then
+			echo -e "On Synology devices, you need to add Plex's public key to Package Center. If you have not done so, follow the instructions at \033[0;34mhttps://support.plex.tv/articles/205165858-how-to-add-plex-s-package-signing-public-key-to-synology-nas-package-center/\033[0m"
```
Since `synopkg` seems to return an exit code of 1 for any and all errors, I'd add in a check for `error = [289]` in the output. (returned on STDOUT even though it's an error, and with an exit code of 1 even though it's obviously 289... way to go Synology... 🤦‍♂)

demonbane replied:
Actually, never mind. We'd have to wrap `DISTRO_INSTALL` with logic to capture the output, and then we'd have to monitor both STDOUT and STDERR and preferably split them out when re-printing them again which would just be a mess.

Though I would remove the `-e` and the color codes. Since this error could theoretically get triggered while running as a scheduled task, color codes could end up garbling the output, so it's best to stick to plain text I think. (and that blue color shows up nearly unreadable on my terminal)

demonbane commented on plexupdate.sh:
```diff
@@ -517,6 +520,8 @@ if [ "${AUTOSTART}" = "yes" ]; then
 	elif [ -x /etc/init.d/plexmediaserver ]; then
 		/etc/init.d/plexmediaserver start
+	elif [ "${DISTRO}" = "synology" ]; then
+		synopkg start "Plex Media Server"
```
Don't forget to also update `extras/installer.sh` to enable `AUTOSTART` when installing on Synology.

myellen replied:
do you think AUTODELETE should be enabled by default if the user chooses to AUTOINSTALL?

demonbane replied:
Hmm. I personally prefer to keep the old copies around just as a way to review my update history, but I think any reasonable person (i.e. not me) would probably expect updates to happen without littering up your hard drive, so I think that's a great idea actually.

demonbane commented:
> We could skip that check altogether, but my synology does have an /etc/crontab and /etc/cron.d that should work.

The installer only knows how to put a file into `/etc/cron.daily` right now anyway (though I am planning support for systemd timers at some point), so I don't think mentioning cron would be useful here.

What I'd **love** would be if the installer can just create a Task Scheduler task from the command line, but I'm not having much luck in finding a way to do it. `synoschedtask` seems to only be able to operate on already defined tasks, not set up a new one, and `synowebapi` could almost certainly do it, but I can't find any public API documentation on the Task Scheduler interface.

That would still be my **preferred** solution, but if we can't find a way to do it then we can at least print a link to a (soon to be written) wiki article on how to set up a scheduled task on Synology. (just need to short-circuit the logic at the top of `configure_cron`)

myellen commented:
The GUI task scheduler just adds the tasks to /etc/crontab using synoschedtask. In my crontab `0   6   *   *   *   root    /usr/syno/bin/synoschedtask --run id=16` is the line for plexupdate.

We should be able to add a line to /etc/crontab, link cronwrapper inside of /etc/cron.d, or create a /etc/cron.daily folder to use.

demonbane commented:
Do you know if cron job output will still get sent to the admin? I have a couple of scheduled tasks on my NAS and when they run I get an HTML email from Synology showing the output which leads me to believe that notifications are handled by synoschedtask, not the usual cron mail configuration.

demonbane commented:
Oh, and the system crontab has `MAILTO=""` right at the top, so I guess that answers that question. :) 

Ok, so in that case we can still add the job, we just need to provide a warning that on Synology you won't get any output. Any thoughts on a good flow to use in the installer that'll also be portable to other Linux systems that may not use the cron.daily setup?

myellen commented:
I think if there's no cron.daily, it should check for cron.d. If it finds cron.d, change CRONWRAPPER to /etc/cron.d/plexupdate, then create a crontab file named plexupdate in cron.d with one line that just calls the cronwrapper as a script like `0   6   *   *   *   root    /path/to/plexupdate/cronwrapper`.

So add some logic around [line 235](https://github.com/mrworf/plexupdate/blob/54da6bca99af8eb9375b446801d080c399141389/extras/installer.sh#L253) and around [line 277](https://github.com/mrworf/plexupdate/blob/54da6bca99af8eb9375b446801d080c399141389/extras/installer.sh#L277) to check CRONWRAPPER and switch to something like `echo "0   6   *   *   *   root   ${FULL_PATH}/extras/cronwrapper" > "$CRONWRAPPER"`

demonbane commented on extras/installer.sh:
```diff
@@ -186,6 +186,8 @@ configure_plexupdate() {
 		[ -z "$DISTRO" ] && check_distro
 		if [ "$DISTRO" == "redhat" ]; then
 			AUTOSTART=yes
+		elif [ "$DISTRO" == "synology" ]; then
```
I have a pathological aversion to duplicating code, so I'd write this as `if [ "$DISTRO" == "redhat" -o "$DISTRO" == "synology" ]; then`.

demonbane commented:
I agree with everything except:

> I think if there's no cron.daily, it should check for cron.d.

At the time, using cron.daily was just the simplest way of implementing this, but if we're going to just use cron.d might as well use it by default and get rid of cron.daily altogether.

Since installer.sh is meant to be run multiple times without problems, there should be some code to migrate an old cron.daily setup to cron.d. (for bonus points, grab the cron.daily time in crontab and use that in the cron.d entry. And if `cron.daily/plexupdate` isn't a symlink just assume they've got a custom setup and exit the cron logic without any changes/questions)

myellen commented:
I changed the cron support to use cron.d instead of cron.daily. There's not anything to migrate since it still uses /etc/plexupdate.cron.conf for settings, and that file gets overwritten every time you run installer.sh anyways. But it does remove the old file in cron.daily if its a symlink.

I look for a cron.daily job in crontab for the time, otherwise I just picked 4 am. May be something we prompt the user for.

I'm having some weird trouble with it running though. The cron job runs at the right time, and it does run the cronwrapper, and that does run plexupdate.sh, but it doesn't actually update. However if I manually run cronwrapper, it works fine. Maybe something with permissions or something weird with cron. So I'm curious if it's just my machine it happens on.

myellen commented:
On second look, it might be because I don't include PATH in the new cron file...

myellen commented:
yeah, that fixed it. Should be all good now.

deepfriedmind commented:
So, is this about to be merged, or..?

demonbane commented:
Hopefully late next week. I’ve been traveling the last few weeks and haven’t had a chance to review the latest updates, but it’s on my list as soon as I get back.

vchatela commented:
Hello all,

I have installed from this PR on my Syno DS216+II. 

I don't receive any mail but I think I have configured well : 
ash-4.3# more /etc/cron.d/plexupdate
MAILTO="chatelard.valentin@gmail.com"
PATH=/sbin:/bin:/usr/sbin:/usr/bin:/usr/syno/sbin:/usr/syno/bin:/usr/local/sbin:/usr/local/bin
#minute hour    mday    month   wday    who     command
0       4       *       *       *       root    /bin/sh /volume1/tools/plex/plexupdate/extras/cronwrapper

Could you please help me ?

deepfriedmind commented:
I've checked out this pull request and it still seems that PMS isn't started again after an update.

myellen commented:
@deepfriedmind Which Synology and DSM version are you using? It’s been working for me on a ds916+ and dsm 6.

deepfriedmind commented:
I'm using a DS916+ on DSM 6.2.2-24922.

myellen commented:
Here are my troubleshooting questions:
Is `AUTOSTART=‘yes’` In your configuration file?
Are `DISTRO` and/or `DISTRO_INSTALL` already defined in your configuration file?
Does the file `/etc/synoinfo.conf` exist on your Synology?
What happens if you use ssh and run `synopkg start “Plex Media Server”`?
Does `synopokg list --name` show a name for plex different form “Plex Media Server”?

gapple commented:
I was able to install and configure this branch successfully on a DS418 with DSM 6.2.3-25426 Update 2

gapple commented:
This no longer appears to work on DSM 7; it looks like there is a new `DISTRO`: `synology-dsm7`
